### PR TITLE
fix #8231 chore(nimbus): use ruff for python linting

### LINKED ISSUE #8231: Replace flake8 with ruff
There's a new linter on the block, and it's called [ruff](https://github.com/charliermarsh/ruff).  It has a lot of nice rules built in, AND it automatically fixes things, so let's replace flake8 with ruff.

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/EXP-3116)

### PR DESCRIPTION
**Pull request by jaredlockhart**
Because

* There is a fancy new linter for Python called ruff
* It has many built in rules from other popular Python linters
* It also applies code fixes for some rules so you get some fancy autoformatting features
* It's also very fast

This commit

* Removes flake8
* Adds ruff
* Enables many of the builtin ruff rules
* Applies ruff fixes

### CODE REVIEW

File: app/experimenter/base/tests/test_base.py [hunk 1]
```diff
@@ -14,43 +14,40 @@ def setUp(self):
...
         expected_version = "8675309"
         version_json = f'{{"commit": "{expected_version}"}}'
-        with self.settings(APP_VERSION=None):
-            with mock.patch(
-                "experimenter.base.open",
-                mock.mock_open(read_data=version_json),
-            ) as mf:
-                version = app_version()
-                self.assertEqual(version, expected_version)
+        with self.settings(APP_VERSION=None), mock.patch(
```
Comment (yashikakhurana) [inline app/experimenter/base/tests/test_base.py:17]: impressed with this formatting and optimisation 🎉

File: app/experimenter/experiments/api/v6/serializers.py [hunk 2]
```diff
@@ -51,12 +52,11 @@ def get_feature(self, obj):
                     fv for fv in feature_values if fv.feature_config == feature_config
                 ][0]
                 feature_enabled = branch_feature_value.enabled
-                try:
-                    feature_value = json.loads(branch_feature_value.value)
-                except json.JSONDecodeError:
+
+                with contextlib.suppress(json.JSONDecodeError):
```
Comment (yashikakhurana) [inline app/experimenter/experiments/api/v6/serializers.py:56]: interesting

File: app/pyproject.toml [hunk 3]
```diff
@@ -118,3 +116,33 @@ reportUntypedBaseClass = false
...
+    "N815",
+    "RET503",
+    "RET504",
+    "RET505",
+    "SIM102",
+]
+
+# Exclude a variety of commonly ignored directories.
+exclude = [
+    "migrations",
```
Comment (yashikakhurana) [inline app/pyproject.toml:141]: previously migrations were not excluded, do we want to keep that?

**Review by yashikakhurana — APPROVED** [review]
Thanks @jaredlockhart some interesting changes syntax, looks good just have one question

File: app/experimenter/experiments/api/v6/serializers.py (hunk 2 above)
Comment (jaredlockhart) [reply app/experimenter/experiments/api/v6/serializers.py:56]: I know I've never seen this pattern before either.  It's very nice.

File: app/pyproject.toml (hunk 3 above)
Comment (jaredlockhart) [reply app/pyproject.toml:141]: Looks like they were: https://github.com/mozilla/experimenter/blob/main/app/.flake8#L2

**Review by eliserichards — APPROVED** [review]
🐶 👍

File: app/experimenter/experiments/api/v6/serializers.py (hunk 2 above)
Comment (jrbenny35) [reply app/experimenter/experiments/api/v6/serializers.py:56]: What the heck this is amazing!

File: app/experimenter/experiments/models.py [hunk 4]
```diff
@@ -801,9 +796,9 @@ def clone(self, to_experiment):
 
 # Helper to ensure branch screenshot filenames have controlled unique paths
 def nimbus_branch_screenshot_upload_to(screenshot, filename):
-    id = uuid4()
+    screenshot_id = uuid4()
     ext = filename.split(".")[-1].lower()
-    return os.path.join(screenshot.branch.experiment.slug, f"{id}.{ext}")
+    return os.path.join(screenshot.branch.experiment.slug, f"{screenshot_id}.{ext}")
```
Comment (jrbenny35) [inline app/experimenter/experiments/models.py:801]: this is some SKYNET stuff

File: app/experimenter/experiments/tests/api/v5/test_queries.py [hunk 5]
```diff
@@ -2072,7 +2072,7 @@ def assertChoices(data, text_choices):
             ]
             self.assertEqual(
                 set(channels),
-                {channel.name for channel in application_config.channel_app_id.keys()},
+                {channel.name for channel in application_config.channel_app_id},
```
Comment (jrbenny35) [inline app/experimenter/experiments/tests/api/v5/test_queries.py:2075]: smart!

**Review by jrbenny35 — COMMENTED** [review]
Some cool stuff for sure

**Pull request merged**